packages/flutter/lib/src/material/about.dart

```diff
@@ -623,3 +982,651 @@ Widget _defaultApplicationIcon(BuildContext context) {
+    assert(() {
+      if (pageOpener == null && !nullOk) {
+        throw FlutterError(
+            'Master Detail operation requested with a context that does not include a Master Detail'
+                ' Flow.\nThe context used to open a detail page from the Master Detail Flow must be'
+                ' that of a widget that is a descendant of a Master Detail Flow widget.');
+      }
+      return true;
+    }());
+    return pageOpener != null ? _MasterDetailFlowProxy._(pageOpener) : null;
+  }
+}
+
+/// Interface for interacting with the [_MasterDetailFlow].
+class _MasterDetailFlowProxy implements _PageOpener {
+  _MasterDetailFlowProxy._(this._pageOpener);
+
+  final _PageOpener _pageOpener;
+
+  /// Open detail page with arguments.
+  @override
+  void openDetailPage(Object arguments) =>
+      _pageOpener.openDetailPage(arguments);
+
+  /// Set the initial page to be open for the lateral layout. This can be set at any time, but
+  /// will have no effect after any calls to openDetailPage.
+  @override
+  void setInitialDetailPage(Object arguments) =>
+      _pageOpener.setInitialDetailPage(arguments);
+}
+
+abstract class _PageOpener {
+  void openDetailPage(Object arguments);
+
+  void setInitialDetailPage(Object arguments);
+}
+
+class _MasterDetailFlowState extends State<_MasterDetailFlow>
+    implements _PageOpener {
+  /// Tracks whether focus is on the detail or master views. Determines behaviour when switching
+  /// from lateral to nested navigation.
+  _Focus focus = _Focus.master;
+
+  /// Cache of arguments passed when opening a detail page. Used when rebuilding.
+  Object _cachedDetailArguments;
+
+  /// Record of the layout that was built.
+  _LayoutMode _builtLayout;
+
+  /// Key to access navigator in the nested layout.
+  final GlobalKey<NavigatorState> _navigatorKey = GlobalKey<NavigatorState>();
+
+  @override
+  void openDetailPage(Object arguments) {
+    _cachedDetailArguments = arguments;
+    if (_builtLayout == _LayoutMode.nested) {
+      _navigatorKey.currentState.pushNamed(_navDetail, arguments: arguments);
+    } else {
+      focus = _Focus.detail;
+    }
+  }
+
+  @override
+  void setInitialDetailPage(Object arguments) {
+    _cachedDetailArguments = arguments;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    switch (widget.displayMode) {
+      case _LayoutMode.nested:
+        return _nestedUI(context);
+      case _LayoutMode.lateral:
+        return _lateralUI(context);
+      case _LayoutMode.auto:
+      default:
+        return LayoutBuilder(
+            builder: (BuildContext context, BoxConstraints constraints) {
+              final double availableWidth = constraints.maxWidth;
+              if (availableWidth >= (widget.breakpoint ?? 840)) {
+                return _lateralUI(context);
+              } else {
+                return _nestedUI(context);
+              }
+            });
+    }
+  }
+
+  Widget _nestedUI(BuildContext context) {
```

Comment:
Consider breaking out this function, and `_lateralUI` into separate `StatelessWidget`s.